src/ctap/apdu.rs

```diff
@@ -0,0 +1,303 @@
+                if apdu.le == 0x00 {
+                    apdu.le = 0x100;
+                }
+            }
+        }
+        // TODO: Add extended length cases
+        if apdu.case_type == 0x00 {
+            return Err(ApduStatusCode::SW_COND_USE_NOT_SATISFIED);
+        }
+        Ok(apdu)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    fn pass_frame(frame: &[u8]) -> Result<APDU, ApduStatusCode> {
+        APDU::try_from(frame)
+    }
+
+    #[test]
+    fn test_case_type_1() {
+        let frame: [u8; 4] = [0x00, 0x12, 0x00, 0x80];
+        let response = pass_frame(&frame);
+        assert!(response.is_ok());
+        let expected = APDU {
+            header: ApduHeader {
+                cla: 0x00,
+                ins: 0x12,
+                p1: 0x00,
+                p2: 0x80,
+            },
+            lc: 0x00,
+            data: Vec::new(),
+            le: 0x00,
+            case_type: 0x01,
+        };
+        assert_eq!(expected, response.unwrap());
+    }
+
+    #[test]
+    fn test_case_type_2_short() {
+        let frame: [u8; 5] = [0x00, 0xb0, 0x00, 0x00, 0x0f];
+        let response = pass_frame(&frame);
+        assert!(response.is_ok());
+        let expected = APDU {
+            header: ApduHeader {
+                cla: 0x00,
+                ins: 0xb0,
+                p1: 0x00,
+                p2: 0x00,
+            },
+            lc: 0x00,
+            data: Vec::new(),
+            le: 0x0f,
+            case_type: 0x02,
+        };
+        assert_eq!(expected, response.unwrap());
+    }
+
+    #[test]
+    fn test_case_type_2_short_le() {
+        let frame: [u8; 5] = [0x00, 0xb0, 0x00, 0x00, 0x00];
+        let response = pass_frame(&frame);
+        assert!(response.is_ok());
+        let expected = APDU {
+            header: ApduHeader {
+                cla: 0x00,
+                ins: 0xb0,
+                p1: 0x00,
+                p2: 0x00,
+            },
+            lc: 0x00,
+            data: Vec::new(),
+            le: 0x100,
+            case_type: 0x02,
+        };
+        assert_eq!(expected, response.unwrap());
+    }
+
+    #[test]
+    fn test_case_type_3_short() {
+        let frame: [u8; 7] = [0x00, 0xa4, 0x00, 0x0c, 0x02, 0xe1, 0x04];
+        let payload = [0xe1, 0x04];
+        let response = pass_frame(&frame);
+        assert!(response.is_ok());
+        let expected = APDU {
+            header: ApduHeader {
+                cla: 0x00,
+                ins: 0xa4,
+                p1: 0x00,
+                p2: 0x0c,
+            },
+            lc: 0x02,
+            data: payload.to_vec(),
+            le: 0x00,
+            case_type: 0x03,
+        };
+        assert_eq!(expected, response.unwrap());
+    }
+
+    #[test]
+    fn test_case_type_4_short() {
+        let frame: [u8; 13] = [
+            0x00, 0xa4, 0x04, 0x00, 0x07, 0xd2, 0x76, 0x00, 0x00, 0x85, 0x01, 0x01, 0xff,
+        ];
+        let payload = [0xd2, 0x76, 0x00, 0x00, 0x85, 0x01, 0x01];
```

Comment:
I would rather compute `payload` as a slice of `frame` here. Otherwise it's not obvious that `payload` is actually already contained in `frame`. Same thing for the tests below.
Alternatively you may leave `payload` as is and construct `frame` using `Vec`:
```rust
let payload = [0xd2, 0x76, 0x00, 0x00, 0x85, 0x01, 0x01];
let frame = {
  let mut frame = vec![0x00, 0xa4, 0x04, 0x00, 0x07];
  frame.extend(&payload);
  let le = [0x01, 0xff];
  frame.extend(&le);
  frame
};
```